python/ray/util/ray_executor.py

```diff
@@ -0,0 +1,317 @@
+        (see https://docs.ray.io/en/latest/ray-core/package-ref.html#ray-init).
+
+        For example, this will connect to a cluster at the specified address:
+        .. code-block:: python
+
+            RayExecutor(address='192.168.0.123:25001')
+
+        Note: excluding an address will initialise a local Ray cluster.
+    """
+
+    def __init__(
+        self, max_workers: Optional[int] = None, shutdown_ray: bool = True, **kwargs
+    ):
+
+        """
+        Initialise a new RayExecutor instance which distributes tasks over
+        a Ray cluster.
+
+        self._shutdown_ray:
+            If False, the Ray cluster is not destroyed when self.shutdown() is
+            called. This prevents initialising a new cluster every time the
+            executor is used in a `with` context. Futures will be
+            destroyed regardless of the value of `_shutdown_ray`.
+        self._shutdown_lock:
+            This is set to True once self.shutdown() is called. Further task
+            submissions are blocked.
+        self._futures:
+            Futures are aggregated into this list as they are returned.
+        self.__actor_pool:
+            ActorPool which is used by self.map() when max_workers > 1
+        self.__remote_fn:
+            Wrapper around the remote function to be executed by Ray.
+        self.context:
+            Context containing settings and attributes returned after
+            initialising the Ray client.
+        """
+
+        self._shutdown_ray: bool = shutdown_ray
+        self._shutdown_lock: bool = False
+        self._futures: List[Future] = []
+        self.__actor_pool: Optional[ActorPool] = None
+        self.__remote_fn: "Optional[RemoteFunction0]" = None
+        self.context: "Optional[BaseContext]" = None
+
+        # The following is necessary because `@ray.remote` is only available at runtime.
+        if max_workers is None:
+
+            @ray.remote
+            def remote_fn(fn: Callable[[], T]) -> T:
+                return fn()
+
+            self.__remote_fn = remote_fn
+        else:
+            if max_workers < 1:
+                raise ValueError(
+                    f"`max_workers={max_workers}` is given. The argument \
+                    `max_workers` must be >= 1"
+                )
+
+            @ray.remote
+            class ExecutorActor:
+                def __init__(self):
+                    pass
+
+                def actor_function(self, fn: Callable):
+                    return fn()
+
+                def ready(self):
+                    return
+
+            actors = [
+                ExecutorActor.options(  # type: ignore[attr-defined]
+                    name=f"actor-{i}"
+                ).remote()
+                for i in range(max_workers)
+            ]
+            for a in actors:
+                ray.get(a.ready.remote())
+            self.__actor_pool = ray.util.ActorPool(actors)
+
+        self.context = ray.init(ignore_reinit_error=True, **kwargs)
```

Comment:
I think you don't need to call `ray.init()`:

1. it's possible that ray is already initialized so we don't need to do it again.
2. if it's not already initialized, it will be auto called when ray methods are called.

Reply:
I've removed actor pools, which resolves this. Thanks.